prow/tide/tide.go

```diff
@@ -756,6 +762,75 @@ func (c *Controller) takeAction(sp subpool, batchPending, successes, pendings, n
 	return Wait, nil, nil
 }
 
+// changedFilesAgent queries and caches the names of files changed by PRs.
+// Cache entries expire if they are not used during a sync loop.
+type changedFilesAgent struct {
+	ghc         githubClient
+	changeCache map[changeCacheKey][]string
+	// nextChangeCache caches file change info that is relevant this sync for use next sync.
+	// This becomes the new changeCache when prune() is called at the end of each sync.
+	nextChangeCache map[changeCacheKey][]string
+	sync.Mutex
+}
+
+type changeCacheKey struct {
+	org, repo string
+	number    int
+	sha       string
+}
+
+// prChanges gets the files changed by the PR, either from the cache or by
+// querying GitHub.
+func (c *changedFilesAgent) prChanges(pr *PullRequest) ([]string, error) {
+	cacheKey := changeCacheKey{
+		org:    string(pr.Repository.Owner.Login),
+		repo:   string(pr.Repository.Name),
+		number: int(pr.Number),
+		sha:    string(pr.HeadRefOID),
+	}
+
+	changedFiles, ok := c.changeCache[cacheKey]
+	// We don't need to hold the lock while reading from changedFiles because it
+	// is only written from prune() which is only called at the very end of Sync()
+	// at which point no other threads could be using this agent.
+	c.Lock()
+	if ok {
+		c.nextChangeCache[cacheKey] = changedFiles
+		c.Unlock()
+		return changedFiles, nil
+	}
+	if changedFiles, ok = c.nextChangeCache[cacheKey]; ok {
+		c.Unlock()
+		return changedFiles, nil
+	}
+	c.Unlock()
+
+	// We need to query the changes from GitHub.
+	changes, err := c.ghc.GetPullRequestChanges(
+		string(pr.Repository.Owner.Login),
+		string(pr.Repository.Name),
+		int(pr.Number),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error getting PR changes for #%d: %v", int(pr.Number), err)
+	}
+	changedFiles = make([]string, 0, len(changes))
+	for _, change := range changes {
+		changedFiles = append(changedFiles, change.Filename)
+	}
+
+	c.Lock()
+	c.nextChangeCache[cacheKey] = changedFiles
+	c.Unlock()
+	return changedFiles, nil
+}
+
+// prune removes any cached file changes that were not used since the last prune.
+func (c *changedFilesAgent) prune() {
+	c.changeCache = c.nextChangeCache
```

Comment:
I think this should be locking the mutex, it's weird to rely on the caller of the caller not being concurrent..